clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -163,6 +175,17 @@ public void run() {
      */
     void run(long now) {
         Cluster cluster = metadata.fetch();
+
+        if (transactionState.isIdempotenceEnabled()) {
+            while (!transactionState.pidIsSet() && (time.milliseconds() - now) < maxBlockTimeMs) {
+                getPID(requestTimeout);
+            }
+            if (!transactionState.pidIsSet()) {
+                transactionState.setStateInvalid();
+                throw new TimeoutException("Could not get PID within reasonable time.");
```

Comment:
Note that this exception is caught in `run()` above. I'm not sure that's the intention.